packages/multichain/src/handlers/wallet-invokeMethod.ts

```diff
@@ -0,0 +1,120 @@
+import type { Caveat } from '@metamask/permission-controller';
+import { providerErrors, rpcErrors } from '@metamask/rpc-errors';
+import type {
+  Json,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+import { numberToHex } from '@metamask/utils';
+
+import { getSessionScopes } from '../adapters/caip-permission-adapter-session-scopes';
+import type { Caip25CaveatValue } from '../caip25Permission';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from '../caip25Permission';
+import { assertIsInternalScopeString } from '../scope/assert';
+import type { ExternalScopeString } from '../scope/types';
+import { parseScopeString } from '../scope/types';
+
+/**
+ * Handler for the `wallet_invokeMethod` RPC method.
+ *
+ * @param request - The request object.
+ * @param _response - The response object.
+ * @param next - The next middleware function.
+ * @param end - The end function.
+ * @param hooks - The hooks object.
+ * @param hooks.getCaveat - the hook for getting a caveat from a permission for an origin.
+ * @param hooks.findNetworkClientIdByChainId - the hook for finding the networkClientId for a chainId.
+ * @param hooks.getSelectedNetworkClientId - the hook for getting the current globally selected networkClientId.
+ */
+async function walletInvokeMethodHandler(
+  request: JsonRpcRequest & { origin: string },
+  _response: PendingJsonRpcResponse<Json>,
+  next: () => void,
+  end: (error: Error) => void,
+  hooks: {
+    getCaveat: (
+      origin: string,
+      endowmentPermissionName: string,
+      caveatType: string,
+    ) => Caveat<typeof Caip25CaveatType, Caip25CaveatValue>;
+    findNetworkClientIdByChainId: (chainId: string) => string | undefined;
+    getSelectedNetworkClientId: () => string;
+  },
+) {
+  const { scope, request: wrappedRequest } = request.params as {
+    scope: ExternalScopeString;
+    request: JsonRpcRequest;
+  };
+
+  assertIsInternalScopeString(scope);
+
+  let caveat;
+  try {
+    caveat = hooks.getCaveat(
+      request.origin,
+      Caip25EndowmentPermissionName,
+      Caip25CaveatType,
+    );
+  } catch (e) {
+    // noop
+  }
+  if (!caveat?.value?.isMultichainOrigin) {
+    return end(providerErrors.unauthorized());
+  }
```

Comment:
Do we want to get rid of these one way guards?

Reply:
yes I believe we do want to get rid of this now!